src/fw/drivers/imu/lsm6dso/lsm6dso.c

```diff
@@ -0,0 +1,837 @@
+/*
+ * Copyright 2025 Matthew Wardrop
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "drivers/accel.h"
+#include "drivers/i2c.h"
+#include "drivers/exti.h"
+#include "drivers/rtc.h"
+#include "kernel/util/sleep.h"
+#include "system/logging.h"
+#include "util/math.h"
+#include "lsm6dso_reg.h"
+
+#include "lsm6dso.h"
+
+// Forward declaration of private functions defined below public functions
+static int32_t prv_lsm6dso_read(void *handle, uint8_t reg_addr, uint8_t *buffer,
+                                uint16_t read_size);
+static int32_t prv_lsm6dso_write(void *handle, uint8_t reg_addr, const uint8_t *buffer,
+                                 uint16_t write_size);
+static void prv_lsm6dso_mdelay(uint32_t ms);
+static void prv_lsm6dso_init(void);
+static void prv_lsm6dso_chase_target_state(void);
+static void prv_lsm6dso_configure_interrupts(void);
+static void prv_lsm6dso_configure_double_tap(bool enable);
```

Comment:
nit: suggest moving the fifo method declarations to just after `*_configure_interrupts`; both here and in the implementation; to keep everything grouped logically.